dbms/src/Functions/FunctionCustomWeekToSomething.h

```diff
@@ -0,0 +1,157 @@
+                        + ". Should be a date or a date with time",
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            if (!isUInt8(arguments[1].type))
+                throw Exception(
+                    "Function " + getName()
+                        + " supports 1 or 2 or 3 arguments. The 1st argument "
+                          "must be of type Date or DateTime. The 2nd argument (optional) must be "
+                          "a constant UInt8 with week mode. The 3nd argument (optional) must be "
+                          "a constant string with timezone name",
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+        else if (arguments.size() == 3)
+        {
+            if (!isDateOrDateTime(arguments[0].type))
+                throw Exception(
+                    "Illegal type " + arguments[0].type->getName() + " of argument of function " + getName()
+                        + ". Should be a date or a date with time",
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            if (!isUInt8(arguments[1].type))
+                throw Exception(
+                    "Function " + getName()
+                        + " supports 1 or 2 or 3 arguments. The 1st argument "
+                          "must be of type Date or DateTime. The 2nd argument (optional) must be "
+                          "a constant UInt8 with week mode. The 3nd argument (optional) must be "
+                          "a constant string with timezone name",
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            if (!isString(arguments[2].type))
+                throw Exception(
+                    "Function " + getName()
+                        + " supports 1 or 2 or 3 arguments. The 1st argument "
+                          "must be of type Date or DateTime. The 2nd argument (optional) must be "
+                          "a constant UInt8 with week mode. The 3nd argument (optional) must be "
+                          "a constant string with timezone name",
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            if (isDate(arguments[0].type) && std::is_same_v<ToDataType, DataTypeDate>)
+                throw Exception(
+                    "The timezone argument of function " + getName() + " is allowed only when the 1st argument has the type DateTime",
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+        else
+            throw Exception(
+                "Number of arguments for function " + getName() + " doesn't match: passed " + toString(arguments.size())
+                    + ", should be 1 or 2 or 3",
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        /// For DateTime, if time zone is specified, attach it to type.
+        if (std::is_same_v<ToDataType, DataTypeDateTime>)
+            return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, 1, 0));
```

Comment:
It's not very clear. The class template `FunctionCustomWeekToSomething` is used only to implement the `week` and `yearWeek` functions, isn't it? `week()` returns an integer 0..53, `yearWeek` returns a string `yyyyww`. So it seems the return type should be `DataTypeUInt8` or `DataTypeUInt16`, not `DataTypeDateTime`.

Reply:
I'll fix it right away, 
`week `return DataTypeUInt8    `yearWeek `return DataTypeUInt32